test/404s.test.js

```diff
@@ -1772,3 +1772,56 @@ test('setNotFoundHandler should be chaining fastify instance', t => {
 
   t.end()
 })
+
+test('Send 404 when frameworkError calls reply.callNotFound', t => {
+  t.test('Dynamic route', t => {
+    t.plan(4)
+    const fastify = Fastify({
+      logger: true,
+      frameworkErrors: (error, req, reply) => {
+        t.equal(error.message, "'%world' is not a valid url component")
+        return reply.callNotFound()
+      }
+    })
+    fastify.get('/hello/:id', () => t.fail('we should not be here'))
+    fastify.inject({
+      url: '/hello/%world',
+      method: 'GET'
+    }, (err, response) => {
+      t.error(err)
+      t.equal(response.statusCode, 404)
+      t.equal(response.payload, '404 Not Found')
+    })
+  })
+
+  t.end()
+})
+
+test('Send 404 when frameworkError calls reply.callNotFound even with setNotFoundHandler', t => {
+  t.test('Dynamic route', t => {
+    t.plan(4)
+    const fastify = Fastify({
+      logger: true,
+      frameworkErrors: (error, req, reply) => {
+        t.equal(error.message, "'%world' is not a valid url component")
+        return reply.callNotFound()
+      }
+    })
+
+    fastify.setNotFoundHandler(function (_req, reply) {
+      reply.code(404).send('This should not be sent')
```

Comment:
I think this assertion is wrong: I would expect to call this function from the `frameworkErrors`

It should be doable by updating the `onBadUrlContext` on:

```
function setNotFoundHandler (opts, handler) {
```

Reply:
I am sorry I didn't get your comment here. What I understand is that we should call `fastify.setNotFoundHandler` from `frameworkErrors` is it ?

setNotFoundHandler would be accessible post instantiation right ? but frameworkErrors have to be set during instantiation. 

I am bit confused here. Can you shed some light on it ?

Reply:
No, what I mean is that `reply.callNotFound()` "Invokes the custom not found handler" as written in the docs.

In this test, the `reply.callNotFound()` the custom not found handler is ignored by this fix.

I think it is correct to set the `onBadUrlContext[kFourOhFourContext]` to `null`, but the same field can be updated by editing this function:

https://github.com/fastify/fastify/blob/b15c7c566888c7953715c031a8b3bfe13c3d1960/fastify.js#L623

adding a `onBadUrlContext[kFourOhFourContext] = something`

Reply:
What I understood is that `reply.callNotFound()` invokes the preHandler `Invokes the custom not found handler. Note that it will only call 'preHandler' hook specified in ['setNotFoundHandler']` . But unfortunately in our case once we see the `reply.context[kFourOhFourContext] === null` we reply with 404 and the preHandler would never be called.

Reply:
I mean this:

```diff
diff --git a/fastify.js b/fastify.js
index 4a53945b..2507e21a 100644
--- a/fastify.js
+++ b/fastify.js
@@ -610,6 +610,7 @@ function fastify (options) {
 
       childLogger.info({ req }, 'incoming request')
 
+      onBadUrlContext[kFourOhFourContext] = this[kFourOhFourContext]
       const request = new Request(id, null, req, null, childLogger, onBadUrlContext)
       const reply = new Reply(res, request, childLogger)
       return frameworkErrors(new FST_ERR_BAD_URL(path), request, reply)
diff --git a/test/404s.test.js b/test/404s.test.js
index 5b7f13a1..f02764a0 100644
--- a/test/404s.test.js
+++ b/test/404s.test.js
@@ -1809,6 +1809,7 @@ test('Send 404 when frameworkError calls reply.callNotFound even with setNotFoun
     })
 
     fastify.setNotFoundHandler(function (_req, reply) {
+      t.pass('the custom 404 handler is called')
       reply.code(404).send('This should not be sent')
     })
 
@@ -1819,7 +1820,7 @@ test('Send 404 when frameworkError calls reply.callNotFound even with setNotFoun
     }, (err, response) => {
       t.error(err)
       t.equal(response.statusCode, 404)
-      t.equal(response.payload, '404 Not Found')
+      t.equal(response.payload, 'This should not be sent')
     })
   })
 

```